### PR TITLE
[opencascade] fix #16252 

### LINKED ISSUE #16252: Opencascade "Cannot open include file: 'Standard.hxx'"
**Describe the bug**
Opencascade 7.5.0 builds (as windows-x64-static) without any errors. 
However Visual Studio 2019 can not find the header files for Opencascade itself whenever a header from said library is included.
For example when I want to `#include <opencascade/gp.hxx>`, Visual Studio 2019 says:
`1>C:\Sys\vcpkg\installed\x64-windows-static\include\opencascade\gp.hxx(20,10): fatal error C1083: Cannot open include file: 'Standard.hxx': No such file or directory`

**Environment**
- OS: Windows 10
- Compiler: Microsoft (R) C/C++ Optimizing Compiler Version 19.28.29337 for x86

**To Reproduce**
1. ./vcpkg install opencascade.
2. Create empty project (C++) in Visual Studio 2019.
3. Create a .cpp file and `#include <opencascade/gp.hxx>`.
4. See error: `1>C:\Sys\vcpkg\installed\x64-windows-static\include\opencascade\gp.hxx(20,10): fatal error C1083: Cannot open include file: 'Standard.hxx': No such file or directory`.

### PR DESCRIPTION
**Pull request by Dejauxvue**
fix #16252 by exporting include directories with the targets and replacing internal includes from #include <...> to #include "..." post installation

tested on x64-windows

### CODE REVIEW

File: ports/opencascade/portfile.cmake [hunk 1]
```diff
@@ -45,6 +46,30 @@ vcpkg_install_cmake()
 
 vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/opencascade TARGET_PATH share/opencascade)
 
+
```
Comment (NancyLi1013) [inline ports/opencascade/portfile.cmake:49]:
```suggestion
vcpkg_fixup_cmake_targets(CONFIG_PATH lib/cmake/opencascade)
```
`TARGET_PATH ` will specify to `share/${PORT}` by default. So there is no need to add this argument.

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/opencascade/install-include-dir.patch [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
++++ b/CMakeLists.txt
+@@ -1171,6 +1171,15 @@ foreach (OCCT_MODULE ${OCCT_MODULES})
+ endforeach()
+ set (OCCT_MODULES_ENABLED ${OCCT_MODULES_ENABLED_SORTED})
+ 
++foreach (OCCT_LIBRARY ${OCCT_LIBRARIES})
++	target_include_directories(
++		${OCCT_LIBRARY} 
++		INTERFACE
++		#$<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/source>
```
Comment (NancyLi1013) [inline ports/opencascade/install-include-dir.patch:22]: why do you comment this out? Does it not need `BUILD_INTERFACE` here?

File: ports/opencascade/portfile.cmake [hunk 3]
```diff
@@ -45,6 +46,30 @@ vcpkg_install_cmake()
...
+foreach(file_name ${files})
+	file(READ ${file_name} filedata)
+	string(REGEX REPLACE "#include \<([a-zA-Z0-9_]*\.[hgl]xx)\>" "#include \"\\1\"" filedata "${filedata}")
+	foreach(extra_header ${ADDITIONAL_HEADERS})
+		string(REGEX REPLACE "#include \<${extra_header}\>" "#include \"${extra_header}\"" filedata "${filedata}")
+	endforeach()
+	file(WRITE ${file_name} "${filedata}")
+endforeach()
+
+
```
Comment (NancyLi1013) [inline ports/opencascade/portfile.cmake:72]:
```suggestion
```

File: ports/opencascade/portfile.cmake [hunk 4]
```diff
@@ -85,3 +110,5 @@ else()
 endif()
 
 file(INSTALL ${SOURCE_PATH}/OCCT_LGPL_EXCEPTION.txt DESTINATION ${CURRENT_PACKAGES_DIR}/share/${PORT} RENAME copyright)
+
+
```
Comment (NancyLi1013) [inline ports/opencascade/portfile.cmake:114]:
```suggestion
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

File: ports/opencascade/install-include-dir.patch (hunk 2 above)
Comment (Dejauxvue) [reply ports/opencascade/install-include-dir.patch:22]: I wanted to make the changes as small as possible and did not want to touch the include directories for build, since was fine before.
Haven't tested with BUILD_INTERFACE though

Comment (JackBoosY) [reply ports/opencascade/install-include-dir.patch:22]: This will change the upstream behavior, please restore the commented code

File: ports/opencascade/portfile.cmake [hunk 5]
```diff
@@ -43,7 +44,28 @@ vcpkg_configure_cmake(
...
+file(GLOB files "${CURRENT_PACKAGES_DIR}/include/opencascade/[a-zA-Z0-9_]*\.[hgl]xx")
+foreach(file_name ${files})
+	file(READ ${file_name} filedata)
+	string(REGEX REPLACE "#include \<([a-zA-Z0-9_]*\.[hgl]xx)\>" "#include \"\\1\"" filedata "${filedata}")
+	foreach(extra_header ${ADDITIONAL_HEADERS})
+		string(REGEX REPLACE "#include \<${extra_header}\>" "#include \"${extra_header}\"" filedata "${filedata}")
+	endforeach()
+	file(WRITE ${file_name} "${filedata}")
+endforeach()
```
Comment (JackBoosY) [inline ports/opencascade/portfile.cmake:68]: I think we should report this to the upstream first, then use patch to fix that.

**Review by JackBoosY — CHANGES_REQUESTED** [review]

File: ports/opencascade/install-include-dir.patch (hunk 2 above)
Comment (Dejauxvue) [reply ports/opencascade/install-include-dir.patch:22]: the original port didn't have BUILD_INTERFACE 
the comment was left from my personal testing
should I still add it by uncommenting?

Comment (JackBoosY) [reply ports/opencascade/install-include-dir.patch:22]: @Dejauxvue Ahh, so just remove it would be okay, we should reduce the patch size as much as possible.

File: ports/opencascade/portfile.cmake (hunk 5 above)
Comment (Dejauxvue) [reply ports/opencascade/portfile.cmake:68]: Do you mean, we should post an issue on the opencascade repo? and change their CMakeLists.txt using a patch?
I'm not very familiar with some cmake or git related expressions ^^

Comment (JackBoosY) [reply ports/opencascade/portfile.cmake:68]: @Dejauxvue Yes, please report this to [opencascade repo](https://dev.opencascade.org/forums), I will help you to make that patch.

Comment (Dejauxvue) [reply ports/opencascade/portfile.cmake:68]: @JackBoosY Do you know how to report an issue to opencascade? When I try to create a MantisBT account, it says "access denied"...
Also the link on their website (https://dev.opencascade.org/get_involved) doesn't seem to work

Comment (JackBoosY) [reply ports/opencascade/portfile.cmake:68]: @Dejauxvue  Alright, I will help you to fix that tomorrow.

Comment (JackBoosY) [reply ports/opencascade/portfile.cmake:68]: I think we don't need to change this, since these headers are in _include/opencascade/_, we can just apply your patch to fix it.

Comment (Dejauxvue) [reply ports/opencascade/portfile.cmake:68]: So apply the changes to the headers as a patch to opencascades CMakeLists.txt?

Comment (JackBoosY) [reply ports/opencascade/portfile.cmake:68]: @Dejauxvue Just remove these lines would be okay. But it doesn't fix the usage using msvc project.

Comment (Dejauxvue) [reply ports/opencascade/portfile.cmake:68]: yep, so this is more of an issue of msvc import than opencascade port?

Comment (JackBoosY) [reply ports/opencascade/portfile.cmake:68]: For msvc project, we should notice the user to add _include/opencascade_ manually.

**Review by JackBoosY — APPROVED** [review]

**Review by NancyLi1013 — APPROVED** [review]

**Comment by NancyLi1013** [comment]
Hi @Dejauxvue

Would you like to add a patch for this changes?

```
diff --git a/adm/templates/OpenCASCADEConfig.cmake.in b/adm/templates/OpenCASCADEConfig.cmake.in
index 4937103b..cd35e07d 100644
--- a/adm/templates/OpenCASCADEConfig.cmake.in
+++ b/adm/templates/OpenCASCADEConfig.cmake.in
@@ -23,6 +23,7 @@ set (OpenCASCADE_DEVELOPMENT_VERSION "@OCC_VERSION_DEVELOPMENT@")
 # - in Windows style: $INSTALL_DIR/cmake
 get_filename_component (OpenCASCADE_INSTALL_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
 get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
+get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
 if (OpenCASCADE_INSTALL_PREFIX MATCHES "/cmake$")
   get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
 endif()
```
Since it was removed by PR #15997, but it is still a problem in current version. So the patch is required.

**Comment by Dejauxvue** [comment]
> 
> 
> Hi @Dejauxvue
> 
> Would you like to add a patch for this changes?
> 
> ```
> diff --git a/adm/templates/OpenCASCADEConfig.cmake.in b/adm/templates/OpenCASCADEConfig.cmake.in
> index 4937103b..cd35e07d 100644
> --- a/adm/templates/OpenCASCADEConfig.cmake.in
> +++ b/adm/templates/OpenCASCADEConfig.cmake.in
> @@ -23,6 +23,7 @@ set (OpenCASCADE_DEVELOPMENT_VERSION "@OCC_VERSION_DEVELOPMENT@")
>  # - in Windows style: $INSTALL_DIR/cmake
>  get_filename_component (OpenCASCADE_INSTALL_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
>  get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
> +get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
>  if (OpenCASCADE_INSTALL_PREFIX MATCHES "/cmake$")
>    get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
>  endif()
> ```
> 
> Since it was removed by PR #15997, but it is still a problem in current version. So the patch is required.

I tried applying the patch directly in a naive way and it doesn't seem to work.
So I tried to reproduce the changes, but I don't really understand what the patch does.
Does it add the line 
```
get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
```
directly after the same line? Does this even have any effect?

File: ports/opencascade/portfile.cmake [hunk 6]
```diff
@@ -8,6 +8,7 @@ vcpkg_from_github(
     HEAD_REF master
     PATCHES
         fix-pdb-find.patch
+		fix-install-prefix-path.patch
```
Comment (NancyLi1013) [inline ports/opencascade/portfile.cmake:11]:
```suggestion
        fix-pdb-find.patch
        fix-install-prefix-path.patch
```

**Review by NancyLi1013 — CHANGES_REQUESTED** [review]

**Comment by NancyLi1013** [comment]
@Dejauxvue 

You can see the details why the patch is required.
 https://github.com/microsoft/vcpkg/issues/14636

**Comment by socuwn** [comment]
> > Hi @Dejauxvue
> > Would you like to add a patch for this changes?
> > ```
> > diff --git a/adm/templates/OpenCASCADEConfig.cmake.in b/adm/templates/OpenCASCADEConfig.cmake.in
> > index 4937103b..cd35e07d 100644
> > --- a/adm/templates/OpenCASCADEConfig.cmake.in
> > +++ b/adm/templates/OpenCASCADEConfig.cmake.in
> > @@ -23,6 +23,7 @@ set (OpenCASCADE_DEVELOPMENT_VERSION "@OCC_VERSION_DEVELOPMENT@")
> >  # - in Windows style: $INSTALL_DIR/cmake
> >  get_filename_component (OpenCASCADE_INSTALL_PREFIX "${CMAKE_CURRENT_LIST_FILE}" PATH)
> >  get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
> > +get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
> >  if (OpenCASCADE_INSTALL_PREFIX MATCHES "/cmake$")
> >    get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
> >  endif()
> > ```
> > 
> > 
> > Since it was removed by PR #15997, but it is still a problem in current version. So the patch is required.
> 
> I tried applying the patch directly in a naive way and it doesn't seem to work.
> So I tried to reproduce the changes, but I don't really understand what the patch does.
> Does it add the line
> 
> ```
> get_filename_component (OpenCASCADE_INSTALL_PREFIX "${OpenCASCADE_INSTALL_PREFIX}" PATH)
> ```
> 
> directly after the same line? Does this even have any effect?

just to clear up, it basicly does a `cd ..` for the install prefix, causing it to install in the correct prefix path instead of the `shared` subfolder

**Review by ras0219-msft — COMMENTED** [review]
Thanks for the fixes! This all generally looks great.

However, I notice that the PR doesn't currently include the portfile changes to swap `<>`s with `""`s that you mentioned [here](https://github.com/microsoft/vcpkg/issues/16252#issuecomment-781217656). They seem necessary for our MSBuild integration to work.

Could you explain why they aren't needed or add then to this PR?

**Comment by Dejauxvue** [comment]
@ras0219-msft 
I had a conversation with @JackBoosY and they suggested that we should add a hint to msvc user that they should include the directory manually. (I don't know how to link to a conversation, but you can find it in this thread)

**Comment by JackBoosY** [comment]
@ras0219-msft For such ports that export specific include paths in cmake, since we did not provide these paths to msbuild integration after building them, I think that manually adding these paths to the include path of the header file is the only solution.
Because there are too many relative paths included, it is difficult for us to fully consider the included code.

**Review by JackBoosY — APPROVED** [review]

**Comment by Dejauxvue** [comment]
shouldn't we then install the includes from opencascade directly to include instead of include/opencascade ?
If adding include/opencascade to the include paths anyways, I don't think, there's much benefit from encapsulating the headers in the include/opencascade subdirectory

**Comment by JackBoosY** [comment]
@Dejauxvue  We will have a internal discussion.

**Comment by JackBoosY** [comment]
@Dejauxvue  After some discussion, we think that your previous modification to the path in portfile.cmake is correct, please restore it. Sorry.

**Comment by Dejauxvue** [comment]
@JackBoosY 
No problem, I'll do it later or tomorrow.
Btw, is there a way, I can test installation of a package without rebuilding it every time?
It takes me about 1h each time to find a bug in the post-build portion of the port file

**Comment by JackBoosY** [comment]
@Dejauxvue I wish too but not. Sorry.

**Review by ras0219-msft — APPROVED** [review]
This looks great to me, thanks for the hard work here!

Just to double check: you've tested this on x64-windows via CMake; have you tested it via MSBuild? Does it work "out of the box" or is the extra include path required for normal usage?

**Review by JackBoosY — APPROVED** [review]

**Comment by Dejauxvue** [comment]
@ras0219-msft 
Thank you for the heads-up! 
There are indeed two things that should be adressed:
- the exported include directory is include/opencascade for cmake. This means that a project configured by cmake needs to include occt headers by ```#include <gp.hxx>```, while code being configured by importing vcpkg.targets into a Visual Studio project needs to include headers by  ```#include <opencascade/gp.hxx>```.
I suggest to change the exported include directory to include instead of opencascade/include so that  ```#include <opencascade/gp.hxx>``` is always the right way

- Secondly, when configuring with CMake, The compilation works fine, however the linker does cannot resolve all symbols because some targets don't get exported. At this point, I don't know why this happens, but I'll look into it.

In Summary: MSBuild works; CMake uses different include conventions and not all targets get exported

**Comment by ras0219-msft** [comment]
Yep, this looks like the right solution to me, thanks for the continued improvement!

Just to note specifically in case it wasn't mentioned here: the CMake change _adds_ the root `include/` as well, so users can include via either method (prefix with `opencascade/` or direct names).
